[로또 1단계] 채넛 미션 제출합니다.

yrsel opened the pull request:
안녕하세요, 페로로 ! :)
🤖 안드로이드 모바일 크루 채넛 입니다. 리뷰 감사합니다 ! 

# 셀프 체크리스트

<!--
모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
-->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

<!--
학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
-->

1. MVC 패턴에 대한 controller 구조가 잘 이루어져있는지 궁금합니다.
    페어와 함께 생각한 `controller`는 뷰에서 값을 받아 도메인으로 전달하고, 도메인에서 반환 값을 받아 controller로 보내는 역할이라고 생각을 했는데 해당 controller가 제 역할을 잘 하고 있는지 궁금합니다!

2. 로또 번호 생성하는 로직을 LottoMachine 객체에서 진행하는데 랜덤값에 대한 테스트를 어떤 방식으로 진행하면 좋을 지 궁금합니다.
    테스트가 어려운 코드는 리팩토링이 필요한 시점이거나 잘못 설계된 코드라는 의견이 있는데 랜덤값에 대한 테스트에 대한 해법이 궁금합니다.

3. TDD를 진행하며 테스트코드와 도메인 코드의 커밋 시점에 대한 질문 드립니다 !
   페어와 진행하며 Git Commit Prefix 와 커밋 단위에 대해 방법에 대해서 생각해보았습니다.
   1번 방식 : 테스트코드 파일을 먼저 `test` 라는 `prefix` 와 함께 커밋한다. 그 후, 도메인 파일을 `feat` 라는 `prefix` 와 함께 커밋한다.
   2번 방식 : 테스트 파일과 도메인 파일 전부를 `test` 또는 `feat` 라는 `prefix` 와 함께 커밋한다.

## 코드 리뷰 커뮤니케이션

<!--
리뷰어에게 질문이 있다면, PR에 직접 코멘트를 남겨 보세요.  
특정 코드 줄에 의견을 남기면 보다 구체적인 피드백을 받을 수 있습니다.  
-->

📌 [GitHub에서 PR에 댓글 남기는 방법](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request)

## 스크린샷

- 정상 동작 스크린샷
<img width="316" alt="정상 동작" src="https://github.com/user-attachments/assets/2b218736-69bb-42fa-83f2-4f74a1203e7b" />

yrsel committed:
feat: 로또 생성 구현

yrsel committed:
chore: 불필요한 파일 제거 (gitkeep)

yrsel committed:
feat: 당첨번호 예외 구현

yrsel committed:
feat: 보너스 번호 예외 구현

yrsel committed:
feat: 당첨 로또 객체 생성

yrsel committed:
docs: 기능 목록 정의

yrsel committed:
feat: 로또 등수 열거객체 구현

yrsel committed:
feat: 로또 등수 판단 함수 구현

yrsel committed:
feat: 로또 객체 내부에 로또 금액 상수화

yrsel committed:
feat: 로또 구입 금액 단위 예외 처리 구현

yrsel committed:
refactor: 로또 구입 금액 객체 네이밍 변경

- PurchaseAmount.kt -> LottoPurchaseAmount.kt
- PurchaseAmountTest.kt -> LottoPurchaseAmountTest.kt
- 네이밍 변경에 따른 Lotto 내부에 존재하던 금액 상수 위치 변경

yrsel committed:
feat: 로또 구입 금액 범위 예외 처리 구현

yrsel committed:
feat: 로또 구입 금액에 따른 로또 개수 반환 구현

yrsel committed:
docs: 로또 번호에 대한 기능 목록 추가

yrsel committed:
feat: 로또 번호 객체 구현

yrsel committed:
refactor: 로또 번호 객체 생성에 따른 구조 변경

- 전) 로또 번호 타입이 List<Int> 이었던 것을
- 현) List<LottoNumber>로 사용하여 원시타입 포장

yrsel committed:
feat: 로또 생성기에서 로또 객체 생성 함수 구현

yrsel committed:
feat: 로또 구입 금액 입력 함수 구현

yrsel committed:
feat: 로또 당첨 번호 입력 함수 구현

yrsel committed:
feat: 로또 당첨 보너스 번호 입력 함수 구현

yrsel committed:
feat: 입력 객체에서 로또 구입 금액 공백 제거

yrsel committed:
feat: 로또 묶음을 가지는 객체 생성

yrsel committed:
feat: 구입한 로또 개수 출력

yrsel committed:
feat: 입력 받을 값에 대한 가이드 출력

yrsel committed:
feat: 로또 구입 금액 로직 구현

yrsel committed:
feat: 로또 구입 개수 출력 로직 구현

yrsel committed:
feat: 로또 묶음 발행 로직 구현

yrsel committed:
feat: 발행한 로또 번호 출력 로직 구현

yrsel committed:
feat: 입력 받은 값을 통해 당첨 로또 생성 로직 구현

yrsel committed:
style: 모덾 패키지 분리

yrsel committed:
docs: 당첨 통계 기능 목록 추가

yrsel committed:
feat: 로또 등수 별 당첨 개수 찾는 로직 구현

yrsel committed:
feat: 로또 수익률 계산 로직 구현

yrsel committed:
feat: 당첨 통계 출력 로직 구현

yrsel committed:
feat: 구입한 로또에 대한 수익률 출력 로직 구현

yrsel committed:
refactor: 출력값 상수화

yrsel committed:
feat: 로또 숫자 6개 아닐 경우에 대한 예외 로직 추가

yrsel committed:
refactor: 로또 번호 개수 상수 재사용

yrsel committed:
style: 로또 번호 객체 내부의 함수 순서 변경

yrsel committed:
refactor: 당첨 통계 객체 내부의 상수화 및 상수 네이밍 변경

yrsel committed:
refactor: 컨트롤러 내 메서드 네이밍 및 순서 변경

yrsel committed:
test: 당첨 로또 수익률 테스트 추가

yrsel committed:
refactor: 수익률 손해 판단 로직 수정

yrsel committed:
test: 수익률 손해 판단 테스트 추가

yrsel committed:
refactor: 실행 결과에 따라 당첨 번호 네이밍 변경

yrsel committed:
feat: 실행 결과에 따라 줄바꿈 추가

yrsel committed:
test: 로또 번호 정상범위 동작 테스트 추가

yrsel committed:
test: 구입 금액에 대한 정상 입력 테스트 추가

yrsel committed:
test: 당첨 로또에 대한 정상 동작 테스트 추가

yrsel committed:
refactor: 당첨 통계 출력 함수 분리

yrsel committed:
refactor: 에러 메시지가 없을 경우 기본 에러 메시지 출력

yrsel committed:
refactor: 2, 3등 판단하는 로직 수정

yrsel committed:
style: 주석 제거

yrsel committed:
style: 불필요한 엘비스 연산자 제거

yrsel committed:
refactor: 로또 개수만큼 생성하는 로직을 로또 머신으로 이동

yrsel committed:
refactor: 발행한 로또만큼 로또 번호 출력하는 로직을 출력 객체로 이동

yrsel committed:
fix: 출력 값 오탈자 수정

yrsel committed:
docs: 기능 목록 업데이트

yrsel committed:
refactor: 객체 생성 로직을 컨트롤러에서 뷰로 이동

laco-dev commented on src/main/kotlin/lotto/controller/LottoController.kt:
```diff
@@ -0,0 +1,78 @@
+import lotto.view.InputView
+import lotto.view.OutputView
+
+class LottoController(
+    private val inputView: InputView,
+    private val outputView: OutputView,
+    private val lottoMachine: LottoMachine,
+) {
```
> MVC 패턴에 대한 controller 구조가 잘 이루어져있는지 궁금합니다.
페어와 함께 생각한 controller는 뷰에서 값을 받아 도메인으로 전달하고, 도메인에서 반환 값을 받아 controller로 보내는 역할이라고 생각을 했는데 해당 controller가 제 역할을 잘 하고 있는지 궁금합니다!

네, 잘 정리하셨습니다.
디테일한 부분은 이어서 남겨드릴게요.

laco-dev commented on src/main/kotlin/lotto/controller/LottoController.kt:
```diff
@@ -0,0 +1,78 @@
+        processLottoStatistics(lottoStatistics)
+    }
+
+    private fun getPurchaseMoney(): LottoPurchaseAmount =
+        try {
+            outputView.printPurchaseAmountGuide()
+            inputView.readLottoPurchaseAmount()
+        } catch (error: IllegalArgumentException) {
+            outputView.printErrorMessage(error.message)
+            getPurchaseMoney()
+        }
```
Controller에서 검사를 통해 다시 재입력받는 것도 좋은 전략입니다.

한가지 생각해 볼만한 점은 입력을 받는 것과 InputView에서 발생한 오류를 OutputView로 넘겨서 출력하는게 적절한가? 에 대해서인데요.
우리가 입력을 키오스크에서 받는다고 가정해보면 키오스크에 UI가 출력되니까요!?

yrsel replied:
`InputView는 입력에 대한 역할만, OutputView는 출력에 대한 역할을 하도록 해야지` 라는 생각으로 페어와 설계를 진행하였습니다.
그렇다보니 입력을 처리하는 과정에서 발생한 에러일지라도 사용자에게 문제가 발생했음을 알리는 것은 출력의 역할로 생각되어 `OutputView`에서 처리하도록 진행하였습니다.
    
    
키오스크로 생각해본다면 재고가 4개인 상품이 있는데 5개를 선택하려는 상황에 `재고가 부족합니다`라는 문구를 보여주도록 설정되어 있다고 가정해본다면,
5개를 선택(입력)하는 순간 화면에 `재고가 부족합니다`라는 문구가 보여진다고(출력) 생각되었는데 혹시 키오스크에 대한 예시가 어떤 의미인지 방향을 제시해주시면 한 번 더 생각해보겠습니다!

관련 커밋 : [5f7cbcd](https://github.com/woowacourse/kotlin-lotto/pull/108/commits/5f7cbcd78d2a2c71905d8fa1655463c630a98426)

laco-dev replied:
입력, 출력이라는 고정관념을 우선 깨야 할 필요가 있습니다.

정확히 말하면 프로그램에 필요한 입력을 받기 위한 프로세스
프로그램에 필요한 출력을 위한 프로세스

2가지 관점으로 접근하는게 좋겠습니다.
가령 입력을 받기 위해 콘솔에서 입력을 받아서 출력은 GUI 프로그램으로 한다고 해봅시다.
그러면 콘솔에는 "OO를 입력해주세요" 라고 띄워서 바로 입력할 수 있게 도와줍니다.
이를 출력 View라고 할 수 있을까요?

laco-dev replied:
이 내용은 디테일하게 들어가면 결국 InputView, OutputView 구분하는게 모호하게 되기 때문인데요.
제가 강조하고 싶었던 부분은 단순히 IllegalArgumentException으로 재고가 부족하다는 것을 퉁치는것이 문제가 되었던 것 같습니다.

이 내용에는 그렇게 고민 많이 하지 마시고 참고로만 넘어가주세요

yrsel replied:
`화면에 보여지는 것은 출력이다`라는 고정 관념으로 프로그래밍을 진행했던 것 같습니다. 
제시해주신 리뷰 덕분에 새로운 시야로 볼 수 있었습니다, 감사합니다 :)

laco-dev commented on src/main/kotlin/lotto/model/LottoMachine.kt:
```diff
@@ -0,0 +1,21 @@
+package lotto.model
+
+class LottoMachine {
+    fun createLottos(lottoCount: Int): Lottos {
+        val lottoBundle = mutableListOf<Lotto>()
+        repeat(lottoCount) {
+            lottoBundle.add(createLotto())
+        }
+        return Lottos(lottoBundle.toList())
+    }
+
+    private fun createLotto(): Lotto {
+        val lottoNumber =
+            (LottoNumber.MIN_NUMBER..LottoNumber.MAX_NUMBER)
+                .shuffled()
+                .take(Lotto.LOTTO_NUMBERS_COUNT)
+                .map { number -> LottoNumber(number) }
+                .sortedBy { lottoNumber -> lottoNumber.number }
+        return Lotto(lottoNumber)
+    }
```
> 로또 번호 생성하는 로직을 LottoMachine 객체에서 진행하는데 랜덤값에 대한 테스트를 어떤 방식으로 진행하면 좋을 지 궁금합니다.
테스트가 어려운 코드는 리팩토링이 필요한 시점이거나 잘못 설계된 코드라는 의견이 있는데 랜덤값에 대한 테스트에 대한 해법이 궁금합니다.

객체지향 관점으로 접근해본다면, 우리가 하고 싶은 것은 로또 번호를 생성하는 것이에요.
그것이 구체적으로 자동으로 만들어질지, 수동으로 만들어질지 모른다는 점이지요.

이 관계가 인터페이스와 구현체의 관계로 볼 수 있습니다.

yrsel replied:
`LottoMachine`인터페이스와 `LottoMachine`인터페이스를 구현하는 `RandomLottoMachine`클래스로 분리해보았습니다 ! 

관련 커밋 
-  [320ab5](https://github.com/woowacourse/kotlin-lotto/pull/108/commits/320ab5cfa2bd7b62646e41716c3a4952034987bb)  
- [4b44d1](https://github.com/woowacourse/kotlin-lotto/pull/108/commits/4b44d1fa4a659b0decc0525d9e3b54575977c926)

laco-dev commented on src/main/kotlin/lotto/controller/LottoController.kt:
```diff
@@ -0,0 +1,78 @@
+            getPurchaseMoney()
+        }
+
+    private fun getLottoCount(purchaseMoney: LottoPurchaseAmount): Int {
+        val lottoCount = purchaseMoney.getLottoCount()
+        outputView.printLottoCount(lottoCount)
+        return lottoCount
+    }
```
로또 번호를 가져오는데 OutputView에 출력까지 한다는 것을 예측하기 어렵지 않을까요?
함수가 2가지의 역할을 하고 있는건 아닌지 생각해봐야 합니다.

yrsel replied:
로또 개수를 가져오는 부분과 출력하는 부분 2가지 역할을 하고 있어 출력하는 부분을 분리했습니다 ! 

관련 커밋 : [5f7cbc](https://github.com/woowacourse/kotlin-lotto/pull/108/commits/5f7cbcd78d2a2c71905d8fa1655463c630a98426)

laco-dev commented on src/test/kotlin/lotto/LottoTest.kt:
```diff
@@ -0,0 +1,59 @@
+                    LottoNumber(6),
+                ),
+            )
+        val expected = 6
```
흔히 테스트 코드의 가독성을 위해서 여러가지 트릭을 사용하고는 하는데요.
지금처럼 테스트 코드가 보기 불편할 정도라면 Fake Constructor 등의 다양한 방법을 이용하고는 합니다.
https://blog.kotlin-academy.com/item-30-consider-factory-functions-instead-of-constructors-e1c747fc475

```kotlin
val lotto = Lotto(1, 2, 3, 4, 5, 6)
```

yrsel replied:
Fake Constructor 방식을 활용해보려고 했으나 `ktlint`에서 메서드명이 대문자로 시작하는 것을 제한하여 `List<Int>`의 확장함수로 `List<Int>.toLotto()`가 `Lotto`객체를 생성하도록 진행했습니다.
테스트 코드 작성할 때 효율적인 방법을 알려주셔서 감사합니다 :-)

관련 커밋 : [c80ff9](https://github.com/woowacourse/kotlin-lotto/pull/108/commits/c80ff925dfaa3d49438473e0428c4cc0cdf8da7d)

laco-dev commented on src/test/kotlin/lotto/LottoTest.kt:
```diff
@@ -0,0 +1,59 @@
+        val expected = 6
+        val actual = lotto.numbers.distinct().size
+
+        Assertions.assertEquals(expected, actual)
```
또한 테스트 코드에 가급적 "로직"을 포함하지 않는 것이 중요합니다.

우리가 알고 싶은 것은 lotto.numbers.distinct().size가 잘 동작하는지가 중요한게 아닙니다.
Lotto가 잘 생성되었는지만 알고싶은 것이죠

yrsel replied:
이 리뷰를 나중에 확인해서 뒤늦게 반영했습니다.😅

관련 커밋 : [83a4fc](https://github.com/woowacourse/kotlin-lotto/pull/108/commits/83a4fc3b6529c95f77c0f2dd0ee804cd7096d1c4)

laco-dev commented on src/test/kotlin/lotto/RankTest.kt:
```diff
@@ -0,0 +1,34 @@
+import java.util.stream.Stream
+
+class RankTest {
+    @ParameterizedTest
```
ParameterizedTest의 name을 활용하면 이런 테스트 이름으로도 만들어 볼 수 있으려나요?

> {0}개의 번호를 맞추고 보너스 번호 일치 여부가 {1}라면 등수는 {2}다

yrsel replied:
관련 커밋 : [34add1](https://github.com/woowacourse/kotlin-lotto/pull/108/commits/34add19d8270c0cfd7e2f1e989ff3cdd424e97d9)

테스트에 대한 결과를 더 명확하게 예측할 수 있었습니다, 👍

laco-dev commented on src/main/kotlin/lotto/model/LottoNumber.kt:
```diff
@@ -0,0 +1,22 @@
+        return number == other.number
+    }
+
+    override fun hashCode(): Int = number
```
동등성을 보장하기 위해 구현한 것으로 보여요.

하지만 이렇게 만든 인스턴스는 모두 새로운 인스턴스이기 때문에 매우 비효율적인 생성이 될 수 있는데요
```kotlin
LottoNumber(0)
LottoNumber(0)
LottoNumber(0)
LottoNumber(0)
LottoNumber(0)
```

이를 개선하기 위해서는 어떤 방법을 사용할 수 있을까요?

yrsel replied:
직접 재정의 해주기보다는 data class 로 변경하여 재정의된 equals, hashcode 를 활용할 수 있었습니다

관련 커밋 : [e428f8](https://github.com/woowacourse/kotlin-lotto/pull/108/commits/e428f880f7d826845cb1acfc11856655f3a9fb51)

비효율적인 인스턴스 생성은 LottoMachine 인터페이스에서 1 ~ 45 의 LottoNumber 객체를 미리 생성해서 매번 새로운 인스턴스 생성을 방지하였습니다.

관련 커밋 : [320ab5](https://github.com/woowacourse/kotlin-lotto/pull/108/commits/320ab5cfa2bd7b62646e41716c3a4952034987bb)

laco-dev commented on src/main/kotlin/lotto/model/LottoPurchaseAmount.kt:
```diff
@@ -0,0 +1,21 @@
+package lotto.model
+
+class LottoPurchaseAmount(
+    val money: Int,
+) {
+    init {
+        require(money in MIN_LOTTO_PURCHASE_AMOUNT..MAX_LOTTO_PURCHASE_AMOUNT) { ERROR_OUT_OF_RANGE_LOTTO_PURCHASE_AMOUNT }
+        require(money % LOTTO_PRICE == NO_REMAINDER) { ERROR_EXIST_REMAINDER }
+    }
+
+    fun getLottoCount(): Int = money / LOTTO_PRICE
+
+    companion object {
+        private const val MIN_LOTTO_PURCHASE_AMOUNT = 1_000
+        private const val MAX_LOTTO_PURCHASE_AMOUNT = 100_000
+        private const val NO_REMAINDER = 0
+        private const val LOTTO_PRICE = 1_000
+        private const val ERROR_OUT_OF_RANGE_LOTTO_PURCHASE_AMOUNT = "구입 금액은 최소 1,000원 이상 최대 100,000원 이하 입니다."
+        private const val ERROR_EXIST_REMAINDER = "구입 금액은 1,000원 단위입니다."
+    }
+}
```
한편으로는 LottoAmount가 로또 가격 1,000원과 강하게 결합하는게 적절한가? 라는 생각도 들어요
가령 A 상점에서는 로또가 1,000원 B 상점에서는 2,000원에 팔수도 있다고 생각하거든요.

로또 가격 1,000원은 그러면 누가 관리하는게 최선일까..?

yrsel replied:
제시해주신 의견에 동의합니다 !

로또 머신마다 발행하는 금액을 달리하면 좋을 것 같다고 생각하였습니다.
-> `LottoMachine` 구현체에서 관리하게 될 경우 구입금액에 대한 예외(1개도 구입 못하는 경우, 나누어 떨어지지 않은 경우 예외가 발생하는 요구사항 있는 경우)에 대한 유효성 검사가 새로 생성되는 `LottoMachine` 구현체가 생길 경우 매번 새롭게 정의되어 진다고 생각되었습니다.

그래서 다른 방향으로 `getLottoCount()`의 매개변수에 `price: Int`로 설정해 사용자에게 구입하려는 로또의 금액(사용자가 가게를 선택한다는 의미)을 입력할 수 있도록 변경했습니다.

관련 커밋 : [ab604b](https://github.com/woowacourse/kotlin-lotto/pull/108/commits/ab604b9c3276f7347a3d6d26f52cb700f7ef4347)

laco-dev commented on src/main/kotlin/lotto/model/LottoStatistics.kt:
```diff
@@ -0,0 +1,43 @@
+package lotto.model
+
+class LottoStatistics(
+    private val lottos: Lottos,
+    private val winningLotto: WinningLotto,
+    private val purchaseMoney: LottoPurchaseAmount,
+) {
```
로또 Statistics는 당첨로또와의 비교를 통해 당첨 결과를 판단하는 것으로 사용이 됩니다.

그런데 lottoStatistics.lottoStatistics가 매우 어색하게 느껴져요.
당첨을 계산하고 비교하는 것을 누구의 역할로 봐야 하는지에 따라 다르겠습니다.

가령 로또 결과를 만드는 것은 WinniLotto가 스스로 할 역할로 볼 수도 있겠죠?

```kotlin
// WinningLotto
fun match(lotto: Lotto): LottoStatistics { ... }
```

yrsel replied:
저도 코드를 작성하며 클래스명과 필드명이 동일한 부분이 어색하게 느껴졌습니다.

`WinningLotto`에 통계 결과에 대한 물음을 던지는 방식으로 진행하고자 했지만 현재 구조에서 `LottoStatistics`를 생성하기 위해서는 `LottoPurchaseAmount`가 필요한 구조입니다.
그렇다면 `WinningLotto`에서 `LottoStatistics` 객체 생성을 위한 목적으로 `LottoPurchaseAmount`에 대해 알게 되어도 괜찮은가? 라는 의문이 들게 되었습니다.

또한, 기존 `lottoStatistics` 필드를 완성하는 메서드만 `WinningLotto`로 이동하는 것은 아닌가 하는 생각이 들었는데 어떤 구조로 `WinningLotto`에게 물음을 던져야할 지 아직 감을 잡지 못했습니다 🥲

laco-dev replied:
반대로 왜 알면 안되는건가요??

앞서 말씀드린 것 처럼, 클래스가 아무런 기능도 없으면 그건 그냥 값 객체로만 사용되거나 도메인 로직을 다른 누군가가 대신 하고 있다는 뜻일 수 있습니다.

yrsel replied:
객체에 대한 정보는 내부에서 기능을 위해 활용되는 경우에만 참조하여 최소화하는 것이 좋다고 판단 했었습니다.

리뷰를 참고하며 다시 코드를 읽다보니 로또 결과를 만드는 것을 `WinningLotto`의 역할로 보는 것이 더 자연스럽고 알맞은 역할로 보여져 `WinningLotto`객체가 `Lotto`들과 비교 후 생성하도록 코드를 수정했습니다.

관련 커밋 : [181577](https://github.com/woowacourse/kotlin-lotto/pull/108/commits/181577e2f4b475a0494a232c511f6fc524000d87)

laco-dev commented on src/main/kotlin/lotto/model/Lottos.kt:
```diff
@@ -0,0 +1,5 @@
+package lotto.model
+
+class Lottos(
+    val lottoBundle: List<Lotto>,
+)
```
Lottos에 아무런 역할이나 기능이 없다보니 일급 컬렉션의 목적으로 사용이 다소 애매하다고 느껴지네요. 🤔

yrsel replied:
이 부분에 대해서 `List<Lotto>`자료형을 감싸는 용도로만 사용하는 게 맞나 고민 되었습니다.
그래도 `List<Lotto>` 자체를 외부에 노출시키는 것 보다는 감싸주는 것이 낫다고 판단하고 현재 프로그램에서 관리해야 하는 클래스 수가 많지 않기에 `Lottos` 클래스를 만들어 보았습니다.

페로로라면 이런 용도의 일급 컬렉션이라면 사용하지 않는 것을 지향하시나요 ??

laco-dev replied:
아무런 특징도 기능도 없는 클래스라면, 이 객체가 해야할 역할을 다른 녀석이 뺐어가지는 않았는가?
를 우선 고민해볼 필요가 있다고 봅니다.

단순히 Wrapping만 한 것이라면, 다른 개발자가 Lottos 클래스를 사용해야 하는구나 라는 사실 자체를 인지하지 못할 가능성이 있습니다.

그렇다고 이 클래스를 넘겨주도록 하면.. 이거 왜 만들었어요? 라는 생각이 들기도 하겠죠.
말씀주신 내용처럼 모든 내용을 감싸기 위한 연습으로 일급 컬렉션을 적용한 것을 좋습니다.
다만 이렇게 할 것이라면 List 파라미터를 private으로 감싸도록 하고 동일한 구현이 되도록 만들어 보시면 연습이 될 거에요

laco-dev reviewed (changes requested):
반갑습니다~ 로또 미션동안 잘 부탁드립니다.

MVC 패턴을 연습하면서 필요한 기본적인 구조는 잘 작성을 해 주신 것 같습니다.
디테일한 관점에서 몇가지 의견을 드려 봤구요.

설계 구조적으로 고려해볼만한 점이 보여서 수정이 되면 이어서 다른 부분들도 보도록 하겠습니다~

laco-dev commented:
> TDD를 진행하며 테스트코드와 도메인 코드의 커밋 시점에 대한 질문 드립니다 !
페어와 진행하며 Git Commit Prefix 와 커밋 단위에 대해 방법에 대해서 생각해보았습니다.
1번 방식 : 테스트코드 파일을 먼저 test 라는 prefix 와 함께 커밋한다. 그 후, 도메인 파일을 feat 라는 prefix 와 함께 커밋한다.
2번 방식 : 테스트 파일과 도메인 파일 전부를 test 또는 feat 라는 prefix 와 함께 커밋한다.

이거 답변을 안드렸네요.
엄밀히 말하면 하나의 기능은 테스트와 같이 커밋에 포함되어야 합니다.
가령 구현한 기능이 문제가 생겨 수정을 하는 상황에 있어 커밋을 롤백하더라도 테스트가 남아있으면 컴파일 오류가 발생합니다.

유지보수 차원에서 테스트 코드를 추가로 커밋하는 경우는 있지만, TDD 관점에서 새로운 기능을 만드는 것은 항상 세트입니다.

yrsel committed:
refactor: 로또 개수 반환하는 로직 역할 분리

getLottoCount()에서 로또 개수를 반환하는 로직과 출력하는 로직이 섞여 있어서 분리 진행

yrsel committed:
refactor: 로또 등수 판단 테스트 데이터에 대한 이름 추가

맞춘 번호와 보너스 일치 여부, 등수에 대한 정보를 ParameterizedTest의 name 값으로 넣어서 테스트에 대한 예측 결과가 명확하도록 수정

yrsel committed:
refactor: LottoNumber 클래스 변경

equals, hashcode를 직접 재정의해주기 보다는 data class를 활용하여 언어에서 재정의해주는 방향으로 변경

yrsel committed:
feat: 로또 생성기 추상화 객체 구현

로또 생성기를 추상화하였고 1 ~ 45의 LottoNumber를 미리 생성해서 새로운 인스턴스 생성을 방지하여 메모리를 효율적으로 사용하도록 하였습니다

yrsel committed:
refactor: RandomLottoMachine으로 변경

LottoMachine -> RandomLottoMachine 으로 파일,클래스명을 수정하였고 LottoMachine 인터페이스를 구현하도록 변경하였습니다

yrsel committed:
style: 패키지명 변경에 따른 import문 수정

LottoMachine의 패키지를 생성하게 되어 LottoMachine -> lottomachine.LottoMachine 으로 import문을 수정하였습니다

yrsel committed:
refactor: 로또 머신 구현 객체 주입

LottoMachine 을 추상화함으로써 구현 객체인 RandomLottoMachine 을 주입하였습니다

yrsel committed:
refactor: 구입 가능한 로또 개수 반환 로직 수정

사용자가 상점을 선택할 수 있다는 생각을 반영하여 직접 인자를 넣을 수 있도록 수정하였습니다

yrsel committed:
refactor: 테스트용 Lotto 객체 fake 메서드 생성

Lotto 생성에 대한 fake constructor 를 생성하려고 했으나 ktlint 에서 대문자로 시작하는 메서드를 제한해 List<Int>의 확장함수를 활용하였습니다

laco-dev commented on src/main/kotlin/lotto/controller/LottoController.kt:
```diff
@@ -0,0 +1,78 @@
+        return winningLotto
+    }
+
+    private fun getWinningLottoNumbers(): Lotto =
+        try {
+            outputView.printWinningLottoNumbersOfLastWeekGuide()
+            inputView.readWinningLottoNumbersOfLastWeek()
+        } catch (error: IllegalArgumentException) {
+            outputView.printErrorMessage(error.message)
+            getWinningLottoNumbers()
+        }
+
+    private fun createWinningLotto(winningLottoNumbers: Lotto): WinningLotto =
+        try {
+            val bonusNumber = getBonusNumber()
+            WinningLotto(winningLottoNumbers, bonusNumber)
+        } catch (error: IllegalArgumentException) {
+            outputView.printErrorMessage(error.message)
+            createWinningLotto(winningLottoNumbers)
+        }
```
어떤건 get()에서 오류처리하고 어떤건 create()에서 오류 처리하는게 어색하게 느껴져요!

yrsel replied:
`getWinningLotto`가 이미 존재해서 메서드 오버로딩해서 사용해도 괜찮을 까 고민하다가 `createWinningLotto`로 명명했었는데 
코드의 일관성을 지키는 것이 좋을 것 같아 `getWinningLotto` 메서드 오버로딩하는 방식으로 변경했습니다.

관련 커밋 : [5ded75](https://github.com/woowacourse/kotlin-lotto/pull/108/commits/5ded753b13eec935d021db6d19cba50d7ff21927)

laco-dev commented on src/main/kotlin/lotto/controller/LottoController.kt:
```diff
@@ -0,0 +1,78 @@
+        return inputView.readBonusNumber()
+    }
+
+    private fun processLottoStatistics(lottoStatistics: LottoStatistics) {
+        outputView.printLottoStatistics(lottoStatistics.lottoStatistics)
+        val rateOfReturn = lottoStatistics.getRateOfReturn()
+        outputView.printLottoRateOfReturn(rateOfReturn, lottoStatistics.getIsLossMoney(rateOfReturn))
+    }
```
OutputView.printLottoStatistics()에서 한번에 다 출력해도 될텐데 나눠서 호출한 이유는 primitive 타입만 넘기려는 의도인가요?
이미 Rank 넘긴 시점부터 이건 앞뒤가 안맞는 흐름이 되는 것 같아서요..

꼭 이렇게 해야한다는 것은 아니고, 의견이 궁금합니다.

yrsel replied:
수익률을 제외한 당첨 통계만 출력하고 싶은 경우가 있지 않을까? 또는 반대의 경우가 있을 것 같아서 구분해 보았습니다 !
페어와 `Input/OutputView` 에서 객체에 대한 정보를 알아도 괜찮다는 방향으로 설계를 진행 했어서 원시타입만 넘기려는 의도는 아니였습니다 🙂

laco-dev commented on src/main/kotlin/lotto/model/LottoPurchaseAmount.kt:
```diff
@@ -0,0 +1,21 @@
+    companion object {
+        private const val MIN_LOTTO_PURCHASE_AMOUNT = 1_000
+        private const val MAX_LOTTO_PURCHASE_AMOUNT = 100_000
+        private const val NO_REMAINDER = 0
```
개인적으로 개수가 나눠떨어져야 하기 위한 0값 마저 상수화로 만들기엔 조금 과하게 느껴지는 부분이 있네요.

laco-dev commented on src/main/kotlin/lotto/model/LottoStatistics.kt:
```diff
@@ -0,0 +1,43 @@
+        return sum
+    }
+
+    fun getIsLossMoney(rateOfReturn: Double): Boolean = rateOfReturn < 1
```
Kotlin 코드 프로퍼티가 Java로 변환될 때 기본적으로 getter(), setter()를 추가합니다.
기본적으로 getter()는 앞에 getXXX() 형태로 만들어지지만,
Boolean 형태의 isXXX()는 그대로 isXXX() 형태로 만들어집니다.

그렇기에 이 경우는 자연스럽게 get()을 안붙이고는 하는데요~
나중에 시간이 많이 남는다면 kotlin to java 원리를 한번 찾아보세요

yrsel replied:
공부 키워드 감사합니다 ! :)

관련 커밋 : [c3375a](https://github.com/woowacourse/kotlin-lotto/pull/108/commits/c3375a53138b1b6e1731153b95d17edec4cb67b9)

laco-dev commented on src/main/kotlin/lotto/model/LottoStatistics.kt:
```diff
@@ -0,0 +1,43 @@
+        return sum
+    }
+
+    fun getIsLossMoney(rateOfReturn: Double): Boolean = rateOfReturn < 1
```
그런데 이미 Statistics가 rateOrReturn을 알고 있는데 마이너스인지 알기 위해 다시 넘겨주는 이유가 뭘까요?

yrsel replied:
현재 구조에서 `rateOfReturn`은 필드가 아닌 메서드로 연산 후 값을 얻을 수 있는 구조였어서 `getIsLossMoney`에 값을 넣어주지 않는다면 외부에서 여러 번 호출될 때마다 매번 연산이 진행된다고 생각했습니다. 
또한, 테스트가 가능하다는 장점이 있어서 현재 구조를 택했습니다.

당첨 결과에 대한 정보가 변경되는 경우가 없기에 `rateOfReturn`을 필드값으로 설정하고 `getIsLossMoney`메서드의 매개변수를 제거하도록 변경하였습니다.

관련 커밋 : [15d021](https://github.com/woowacourse/kotlin-lotto/pull/108/commits/15d021e0fd624a771b77fb7e240e39d78b1e7db4)

laco-dev commented on src/main/kotlin/lotto/model/WinningLotto.kt:
```diff
@@ -0,0 +1,23 @@
+    val bonusNumber: LottoNumber,
+) {
+    init {
+        require(winningNumbers.numbers.contains(bonusNumber).not()) { ERROR_DUPLICATED_BONUS_NUMBER }
```
특정 번호가 포함되어있는지 여부도 Lotto 객체에게 메시지를 전달하면 되지 않을까요?

yrsel replied:
관련 커밋 : [91e123](https://github.com/woowacourse/kotlin-lotto/pull/108/commits/91e123ce0bd0525a62a8d9cb2ea5323caf1dca1b)

laco-dev commented on src/main/kotlin/lotto/model/WinningLotto.kt:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    fun findLottoRank(lotto: Lotto): Rank {
+        val matchCount =
+            winningNumbers.numbers.count { winningLottoNumber ->
+                lotto.numbers.contains(winningLottoNumber)
+            }
+        val isMatchedBonus = lotto.numbers.contains(bonusNumber)
+        return Rank.valueOf(matchCount, isMatchedBonus)
```
여기도 Lotto 객체가 스스로 "일"을 하도록 만들어보세요

laco-dev commented on src/main/kotlin/lotto/model/lottomachine/LottoMachine.kt:
```diff
@@ -0,0 +1,14 @@
+import lotto.model.Lottos
+
+interface LottoMachine {
+    val lottoNumberBundle: Set<LottoNumber>
+        get() = (LottoNumber.MIN_NUMBER..LottoNumber.MAX_NUMBER).map { number -> LottoNumber(number) }.toSet()
```
머신이 로또 번호를 미리 들고 있더라도 근본적인 문제 해결은 되지 않다고 생각해요.
예를 들어 이렇게 만들면 차이가 없겠죠
```kotlin
val machine1: new LottoMachine
val machine2: new LottoMachine 
val machine3: new LottoMachine 
val machine4: new LottoMachine 
val machine5: new LottoMachine 
```

yrsel replied:
관련 커밋 : [e2ee2c](https://github.com/woowacourse/kotlin-lotto/pull/143/commits/e2ee2cd884fc69b9876888fba5599fe7cf37eb24)

laco-dev commented on src/main/kotlin/lotto/model/lottomachine/LottoMachine.kt:
```diff
@@ -0,0 +1,14 @@
+import lotto.model.Lottos
+
+interface LottoMachine {
+    val lottoNumberBundle: Set<LottoNumber>
+        get() = (LottoNumber.MIN_NUMBER..LottoNumber.MAX_NUMBER).map { number -> LottoNumber(number) }.toSet()
```
추가로 인터페이스에 기본 구현을 넣는건 좋지 않습니다.
인터페이스의 목적 자체가 구현체에게 구현을 위임하는 것이라면 모든 구현체가 반드시 이를 구현해야만 합니다.

로또 이후 미션에서 배울 개념이기 때문에 이정도만 남겨두겠습니다.!!

laco-dev commented on src/main/kotlin/lotto/model/lottomachine/RandomLottoMachine.kt:
```diff
@@ -0,0 +1,20 @@
+import lotto.model.Lottos
+
+class RandomLottoMachine : LottoMachine {
+    override fun createLottos(lottoCount: Int): Lottos {
+        val lottoBundle = mutableListOf<Lotto>()
+        repeat(lottoCount) {
+            lottoBundle.add(createLotto())
+        }
+        return Lottos(lottoBundle.toList())
+    }
```
TIP : List { ... } 빌더 함수를 찾아보세요

yrsel replied:
관련 커밋 : [33e8fd](https://github.com/woowacourse/kotlin-lotto/pull/143/commits/33e8fdf487ccdf30c9145c3f06acb891c4bce704)

laco-dev commented on src/main/kotlin/lotto/view/OutputView.kt:
```diff
@@ -0,0 +1,113 @@
+        rank: Rank,
+        lottoStatistics: Map<Rank, Int>,
+    ) {
+        if (rank == Rank.SECOND) {
+            printLottoSecondRankStatistics(lottoStatistics[rank] ?: INITIAL_LOTTO_RANK_COUNT)
+        } else {
+            printLottoDefaultRankStatistics(rank, lottoStatistics)
+        }
```
보너스 넘버가 있다는 사실을 내부에 숨겼는데 여기서 SECOND로 직접 비교하는게 맞을까요?
만약 3등도 보너스 여부 일치하도록 바뀌어야 하면 OutputView에 있는 이 코드를 실수하지 않을 자신이 있을까요!?

yrsel replied:
보너스 일치여부를 `Rank`에게 물어볼 수 있도록 수정하였습니다.

관련 커밋 : [a14647](https://github.com/woowacourse/kotlin-lotto/pull/108/commits/a14647d71d806d883003c5a7b1e2667d57859e13)

laco-dev commented on src/test/kotlin/lotto/LottoNumberTest.kt:
```diff
@@ -0,0 +1,25 @@
+class LottoNumberTest {
+    @ParameterizedTest
+    @ValueSource(ints = [0, 46, -100])
+    fun `로또 번호는 1~45 사이의 숫자가 아닐 경우 예외를 발생한다`(number: Int) {
+        assertThatIllegalArgumentException()
+            .isThrownBy { LottoNumber(number) }
+            .withMessage("로또 번호는 1 ~ 45 사이의 숫자입니다.")
+    }
```
junit5의 `assertThrwos { ... }` 함수를 사용 해보세요

yrsel replied:
`assertThrows`는  should로 시작하는 메서드들을 제공하여 간결하게 테스트할 수 있다는 장점이 있고`assertThatIllegalArgumentException`은 메서드 체이닝을 통해 should로 시작하는 메서드들과 더불어 다양한 검증을 진행할 수 있다는 장점이 있는 것 같습니다.
`assertThrows` 함수가 코드 읽는 사람에게 더 가독성이 좋고 명확해서 추천해주신 걸 까요 ??
예외 테스트의 경우 주로 어떤 방식을 채택하시는 지도 궁금합니다 ! :)

관련 커밋 : [7b2240](https://github.com/woowacourse/kotlin-lotto/pull/108/commits/7b22409fa3ed5146a7185d37b372440ec596fe90)

laco-dev replied:
위 코드와 완전 동일한 역할을 하기에 간단하게 처리할 수 있는 assertThrows를 제안드린 점입니다.
모르고 계실까봐 남겨드린 내용이에요.

laco-dev commented on src/test/kotlin/lotto/LottoPurchaseAmountTest.kt:
```diff
@@ -0,0 +1,46 @@
+            .withMessage("구입 금액은 1,000원 단위입니다.")
+    }
+
+    @ParameterizedTest
+    @ValueSource(ints = [1_000, 14_000, 100_000, 99_000, 21_000])
+    fun `구입 금액은 최소 1,000원 이상 최대 100,000원 이하이면서 1,000원 단위일 경우 예외를 발생하지 않는다`(money: Int) {
+        Assertions.assertDoesNotThrow {
+            LottoPurchaseAmount(money)
+        }
+    }
```
테스트코드에 예외가 발생한다 발생하지 않는다는 표현은 그다지 좋지는 않습니다.
아직은 비즈니스 오류에 대한 관점이 잘 학습되지 않았지만, 모든 코드를 오류로 잡아내려 하는 것은 서비스 구현 측면에서 좋지는 않기 때문이에요.

이것저것 복잡하게 만들기 보다 단순하게 기능적으로 표현한다면 이렇게 만들어 볼 수도 있겟네요.

> 구입 금액은 1,000원 단위로 나뉘어 떨어져야 한다.

```
assertThrwos { Amount(1,500 }
assertDoesNotThrows { Amount(1,000) }
```

yrsel replied:
모든 코드를 오류로 잡아내면 좋지 않을까 하는 생각이었는데, 서비스 구현 측면에서는 그럴 수도 있겠군요 !

하나의 테스트는 하나에 대해서만(예를 들면, 예외를 발생시킨다 또는 정상 동작한다) 테스트 해야 된다고 생각했습니다.
하나의 테스트 함수 안에 여러 개의 `assert`문이 들어가는 것은 괜찮은 지 궁금합니다.🤔

laco-dev replied:
넵 여러 케이스를 테스트 하는 것은 ParameterizedTest를 사용하지만, 하나의 케이스에서 검증이 필요한 구문이라면 괜찮습니다.

laco-dev commented on src/test/kotlin/lotto/LottoStatisticsTest.kt:
```diff
@@ -0,0 +1,112 @@
+
+    @BeforeEach
+    fun setUp() {
+        lottos =
+            Lottos(
+                listOf(
+                    listOf(1, 2, 3, 4, 5, 6).toLotto(),
+                    listOf(1, 2, 3, 4, 5, 7).toLotto(),
+                    listOf(1, 2, 3, 4, 5, 8).toLotto(),
+                ),
+            )
```
toLotto()랑 비슷하게 더 가독성을 올려보면 좋을 것 같아요

laco-dev commented on src/test/kotlin/lotto/WinningLottoTest.kt:
```diff
@@ -0,0 +1,56 @@
+        val bonusNumber = LottoNumber(7)
+        val winningLotto = WinningLotto(winningNumber, bonusNumber)
+
+        assertThat(winningLotto.winningNumbers).isInstanceOf(Lotto::class.java)
+        assertThat(winningLotto.bonusNumber).isEqualTo(LottoNumber(7))
```
생성자로 넣는 값이 다를리가 없다라는 가정으로 작성하게 되고,
이미 동일한 타입이 아니면 컴파일 타임에 오류로 잡히는데 이를 테스트로 잡아낼 이유가 있나요?

yrsel replied:
객체가 정상적으로 생성되는 지에 대해 테스트하고 싶어서 작성했는데 당연한 결과를 테스트 하고 있었네요 😅

laco-dev reviewed (changes requested):
객체가 많은 역할을 하거나 다른걸 의존하지 않으려는 것에 사로잡혀서 유연함이 조금 떨어진 것 같아요.

물론 정답이 없는 주제인 만큼, 설계가 틀렸다는 것은 아니지만
객관적으로 이 프로그램을 다른 개발자가 사용하더라도 이해가 잘 되는 구조인가? 라는 관점에서는 잘 공감이 안될 것 같은 부분도 있는 것 같아요.

쉽게 표현하면 VO랑 비즈니스 로직이랑 섞여있는 것 같았습니다!
이 내용은 이해하기 어렵기 때문에 너무 머리 싸매지 마시고 이후 미션들에서 익히면 되니 컨디션에 맞춰서 지금 흐름 잘 유지해주시면 좋겠습니다.

다음 요청에는 1단계 마무리할게요 :)

yrsel committed:
refactor: 로또 번호의 중복여부를 로또 객체에게 물어보도록 변경
WinningLotto에서 로또 번호 중복여부를 판단하지 않고 Lotto 객체에게 질문하도록 변경

yrsel committed:
refactor: 로또 번호 일치 개수를 로또 객체에게 물어보도록 변경

yrsel committed:
refactor: WinningLotto가 직접 비교 후 통계 반환

yrsel committed:
test: Rank 관련 테스트 코드 제거

yrsel committed:
test: 도메인 코드 변경에 따른 테스트 코드 변경

yrsel committed:
test: 로또 생성에 대한 테스트 코드 변경

yrsel committed:
refactor: createWinningLotto 메서드명 수정

createWinningLotto -> getWinningLotto(Lotto,LottoPurchaseAmount)

yrsel committed:
refactor: 수익률이 손해인 경우를 판단하는 메서드명 수정

getIsLossMoney -> isLossMoney

yrsel committed:
refactor: 수익률 계산 로직 수정

yrsel committed:
refactor: 보너스볼 일치여부를 Rank 에게 물어보도록 수정

yrsel committed:
refactor: assertThrows{} 로 테스트 코드 수정

yrsel commented:
안녕하세요, 페로로 !
리뷰 내용 잘 봤습니다 :)
페로로의 리뷰를 이해하면서 리팩터링 진행했는데 제시한 올바른 방향으로 진행되고 있는지 모르겠네요 😭
그래도 최대한 반영해보며 저의 주관을 만들어나가려고 노력하고 있습니다 !
전달력이 좋지 않아서 남긴 답글에 대해 이해하기 어려울 수도 있지만 이번에도 잘 부탁드립니다, 감사합니다 ! :)

laco-dev commented on src/main/kotlin/lotto/model/WinningLotto.kt:
```diff
@@ -0,0 +1,30 @@
+package lotto.model
+
+class WinningLotto(
+    val winningNumbers: Lotto,
+    val bonusNumber: LottoNumber,
+    private val purchaseMoney: LottoPurchaseAmount,
+) {
```
수익률을 여러번 계산하는 문제 때문에 WinningLotto에 구매 금액이 들어간 것 같아요.
다만 저는 당첨 로또라는 객체 생성자에 금액이 들어가는게 더 어색하다고 느껴지기는 합니다.

이 값이 통계값을 만드는데만 사용된다면, 함수의 파라미터로 넘겨주는 방법도 괜찮아 보입니다.
```kotlin
fun calculateStatistics(lottos: Lottos, purchaseMoney: PurchaseMoney): LottoStatistics { ... }
```

yrsel replied:
관련 커밋 : [4c66cb](https://github.com/woowacourse/kotlin-lotto/pull/143/commits/4c66cbd790ed13d8cdc5fa9bb2c12e078758cf5a)

laco-dev commented on src/main/kotlin/lotto/view/OutputView.kt:
```diff
@@ -0,0 +1,113 @@
+    fun printLottoStatistics(lottoStatistics: Map<Rank, Int>) {
+        println()
+        println(LOTTO_STATISTICS_TITLE)
+        Rank.entries.reversed().drop(EXCLUDE_MISS_LANK).forEach { rank ->
+            printRankStatistics(rank, lottoStatistics)
+        }
```
MISS를 제외한다는 니즈라면, filter 함수를 활용할 수 있습니다.

yrsel replied:
`drop().forEach{...}`가 1번만 순회한다고 착각했습니다. 
MISS를 제외한다는 로직을 담는 filter 함수를 활용하는 것 방향으로 수정하였습니다 !

yrsel replied:
관련 커밋 : [531838](https://github.com/woowacourse/kotlin-lotto/pull/143/commits/531838a83e3466742e3202be02cb7401782122e2)

laco-dev commented on src/test/kotlin/lotto/LottoStatisticsTest.kt:
```diff
@@ -0,0 +1,50 @@
+import java.util.stream.Stream
+
+class LottoStatisticsTest {
+    private lateinit var purchaseAmount: LottoPurchaseAmount
+
+    @BeforeEach
+    fun setUp() {
+        purchaseAmount = LottoPurchaseAmount(3_000)
+    }
```
PurchaseAmount를 setup()에서 매번 초기화 할 필요가 없을 것 같네요!

yrsel replied:
관련 커밋 : [82ae5b](https://github.com/woowacourse/kotlin-lotto/pull/143/commits/82ae5bfc65245dc8e8529e6009375711bcea852e)

laco-dev reviewed (approved):
네, 고생 많으셨습니다.

깃허브에 숨겨진 리뷰가 몇개 더 남아있으니 이것도 참고해서 같이 챙겨주세요 :)
그럼 2단계 미션도 화이팅입니다.
